toucan_connectors/one_drive/one_drive_connector.py

```diff
@@ -0,0 +1,103 @@
+    ConnectorSecretsForm,
+    ToucanConnector,
+    ToucanDataSource,
+)
+
+API_BASE_ROUTE: str = 'https://graph.microsoft.com/v1.0/me'
+AUTHORIZATION_URL: str = 'https://login.microsoftonline.com/common/oauth2/v2.0/authorize'
+TOKEN_URL: str = 'https://login.microsoftonline.com/common/oauth2/v2.0/token'
+SCOPE: str = 'offline_access Files.Read'
+
+
+class OneDriveDataSource(ToucanDataSource):
+    file: str
+    sheet: str
+    range: Optional[str]
+
+
+class OneDriveConnector(ToucanConnector):
+
+    data_source_model: OneDriveDataSource
+
+    _auth_flow = 'oauth2'
+    _oauth_trigger = 'instance'
+    oauth2_version = Field('1', **{'ui.hidden': True})
+    auth_flow_id: Optional[str]
+
+    def __init__(self, **kwargs):
+        super().__init__(
+            **{k: v for k, v in kwargs.items() if k not in OAuth2Connector.init_params}
+        )
+        # we use __dict__ so that pydantic does not complain about the _oauth2_connector field
+        self.__dict__['_oauth2_connector'] = OAuth2Connector(
+            auth_flow_id=self.auth_flow_id,
+            authorization_url=AUTHORIZATION_URL,
+            scope=SCOPE,
+            token_url=TOKEN_URL,
+            redirect_uri=kwargs['redirect_uri'],
+            config=OAuth2ConnectorConfig(
+                client_id=kwargs['client_id'],
+                client_secret=kwargs['client_secret'],
+            ),
+            secrets_keeper=kwargs['secrets_keeper'],
+        )
+
+    @staticmethod
+    def get_connector_secrets_form() -> ConnectorSecretsForm:
+        return ConnectorSecretsForm(
+            documentation_md=(Path(os.path.dirname(__file__)) / 'doc.md').read_text(),
+            secrets_schema=OAuth2ConnectorConfig.schema(),
+        )
+
+    def build_authorization_url(self, **kwargs):
+        return self.__dict__['_oauth2_connector'].build_authorization_url(**kwargs)
+
+    def retrieve_tokens(self, authorization_response: str):
+        return self.__dict__['_oauth2_connector'].retrieve_tokens(authorization_response)
+
+    def _get_access_token(self):
+        return self.__dict__['_oauth2_connector'].get_access_token()
+
+    def _format_url(self, data_source):
+        url = f'{API_BASE_ROUTE}/drive/root:/{data_source.file}:/workbook/worksheets/{data_source.sheet}/'
+
+        if data_source.range is None:
+            url = url + 'usedRange(valuesOnly=true)'
+        else:
+            url = url + f"range(address='{data_source.range}')"
+
+        return url
+
+    def _run_fetch(self, url):
+        access_token = self._get_access_token()
+        headers = {'Authorization': f'Bearer {access_token}'}
+
+        response = requests.get(url, headers=headers)
+        response.raise_for_status()
+
+        return response.json()
+
+    def _retrieve_data(self, data_source: OneDriveDataSource) -> pd.DataFrame:
+
+        url = self._format_url(data_source)
+
+        response = self._run_fetch(url)
+
+        data = response.get('values')
```

Comment:
I suggest we add a log for empty response. 
`if not data:
    logger.info('No data retrieved from response')`